subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_usb.c

```diff
@@ -0,0 +1,674 @@
+	/*
+	 * Host commands not enabled.
+	 */
+	USB_EC_HOST_CMD_STATE_DISABLED,
+
+	/*
+	 * USB interface is enabled and ready to receive host request.
+	 * Once the response is sent, the current state is reset to this state to accept a next
+	 * request.
+	 */
+	USB_EC_HOST_CMD_STATE_READY_TO_RX,
+
+	/*
+	 * Receiving ongoing. The first part of the host command request has been received,
+	 * potentialy waiting for the rest.
+	 */
+	USB_EC_HOST_CMD_STATE_RECEIVING,
+
+	/*
+	 * The host command request has been fully received and the command is being proccessed.
+	 * The host command handler always has to send a respond, even if the request is invalid.
+	 */
+	USB_EC_HOST_CMD_STATE_PROCESSING,
+
+	/*
+	 * Processing is finished, the response is being sent.
+	 */
+	USB_EC_HOST_CMD_STATE_SENDING,
+};
+
+enum {
+	EC_HOST_CMD_CLASS_ENABLED,
+};
+
+static const char *const state_name[] = {
+	[USB_EC_HOST_CMD_STATE_DISABLED] = "DISABLED",
+	[USB_EC_HOST_CMD_STATE_READY_TO_RX] = "READY_TO_RX",
+	[USB_EC_HOST_CMD_STATE_RECEIVING] = "RECEIVING",
+	[USB_EC_HOST_CMD_STATE_PROCESSING] = "PROCESSING",
+	[USB_EC_HOST_CMD_STATE_SENDING] = "SENDING",
+};
+
+struct ec_host_cmd_desc {
+	struct usb_if_descriptor if0;
+	struct usb_ep_descriptor out_ep;
+	struct usb_ep_descriptor in_bulk_ep;
+	struct usb_ep_descriptor in_int_ep;
+	struct usb_desc_header nil_desc;
+};
+
+struct ec_host_cmd_usb_ctx {
+	struct usbd_class_data *c_data;
+	struct ec_host_cmd_desc *const desc;
+	const struct usb_desc_header **const fs_desc;
+	struct ec_host_cmd_rx_ctx *rx_ctx;
+	struct ec_host_cmd_tx_buf *tx_buf;
+	uint8_t *bulk_out_buf;
+	struct net_buf *usb_rx_buf;
+	struct net_buf *usb_tx_buf;
+	enum ec_host_cmd_usb_state state;
+	ATOMIC_DEFINE(class_state, 1);
+	bool pending_event;
+	struct k_work_delayable reset_work;
+};
+
+static int expected_request_len(const struct ec_host_cmd_request_header *header)
+{
+	/* Check host request version */
+	if (header->prtcl_ver != EC_HOST_REQUEST_VERSION) {
+		return 0;
+	}
+
+	/* Reserved byte should be 0 */
+	if (header->reserved) {
+		return 0;
+	}
+
+	return sizeof(*header) + header->data_len;
+}
+
+static inline uint8_t ec_host_cmd_get_out_ep(struct usbd_class_data *const c_data)
+{
+	struct ec_host_cmd_usb_ctx *ctx = usbd_class_get_private(c_data);
+	struct ec_host_cmd_desc *desc = ctx->desc;
+
+	return desc->out_ep.bEndpointAddress;
+}
+
+static inline uint8_t ec_host_cmd_get_in_bulk_ep(struct usbd_class_data *const c_data)
+{
+	struct ec_host_cmd_usb_ctx *ctx = usbd_class_get_private(c_data);
+	struct ec_host_cmd_desc *desc = ctx->desc;
+
+	return desc->in_bulk_ep.bEndpointAddress;
+}
+
+static inline uint8_t ec_host_cmd_get_in_int_ep(struct usbd_class_data *const c_data)
+{
+	struct ec_host_cmd_usb_ctx *ctx = usbd_class_get_private(c_data);
+	struct ec_host_cmd_desc *desc = ctx->desc;
+
+	return desc->in_int_ep.bEndpointAddress;
+}
+
+static struct net_buf *ec_host_cmd_buf_alloc(const uint8_t ep, size_t size, void *const data)
```

Comment:
Looks like size parameter can be const as well.

`static struct net_buf *ec_host_cmd_buf_alloc(const uint8_t ep, const size_t size, void *const data)`

Reply:
Done.